### ISSUE TITLE
tty0tty failed after update to Linux 6.1

### ISSUE THREAD

**Issue by Dieter9** [freemed/tty0tty#47]
After Arch Linux got an update to Linux 6.1.1, tty0tty does not compile anymore.

```
make[1]: Entering directory '/home/user/work/src/tty0tty-/module'
make -C /lib/modules/6.1.1-arch1-1/build M=/home/user/work/src/tty0tty-/module modules
make[2]: Entering directory '/usr/lib/modules/6.1.1-arch1-1/build'
  CC [M]  /home/user/work/src/tty0tty-/module/tty0tty.o
/home/user/work/src/tty0tty-/module/tty0tty.c:608:24: error: initialization of ‘void (*)(struct tty_struct *, const struct ktermios *)’ from incompatible pointer type ‘void (*)(struct tty_struct *, struct ktermios *)’ [-Werror=incompatible-pointer-types]
  608 |         .set_termios = tty0tty_set_termios,
      |                        ^~~~~~~~~~~~~~~~~~~
/home/user/work/src/tty0tty-/module/tty0tty.c:608:24: note: (near initialization for ‘serial_ops.set_termios’)
cc1: some warnings being treated as errors
make[3]: *** [scripts/Makefile.build:250: /home/user/work/src/tty0tty-/module/tty0tty.o] Error 1
```

**Comment by Dieter9** [comment]
Here is a patch, that makes tty0tty compile and running again on linux 6.1.

```
diff -ur tty0tty-org/module/tty0tty.c tty0tty-master/module/tty0tty.c
--- tty0tty-org/module/tty0tty.c	2022-12-24 10:24:40.149392717 +0100
+++ tty0tty-master/module/tty0tty.c	2022-12-28 20:57:06.306970029 +0100
@@ -288,7 +288,7 @@
 #define RELEVANT_IFLAG(iflag) ((iflag) & (IGNBRK|BRKINT|IGNPAR|PARMRK|INPCK))
 
 static void tty0tty_set_termios(struct tty_struct *tty,
-				struct ktermios *old_termios)
+				const struct ktermios *old_termios)
 {
 	unsigned int cflag;
 	unsigned int iflag;
```

**Comment by dreirund** [comment]
> Here is a patch, that makes tty0tty compile and running again on linux 6.1.

This breaks compilation for 6.0.x kernels, though. And there are still kernels <6.1 around that are still up to date and maintained, e.g. -LTS kernels.

So a patch (maybe a patch to the build system to selectively apply your patch) that makes it work on older _and_ newer kernels is needed, I suppose.

**Comment by sherpya** [comment]
```
diff --git a/module/tty0tty.c b/module/tty0tty.c
index ce4b73b..c234f85 100644
--- a/module/tty0tty.c
+++ b/module/tty0tty.c
@@ -288,6 +288,9 @@ exit:
 #define RELEVANT_IFLAG(iflag) ((iflag) & (IGNBRK|BRKINT|IGNPAR|PARMRK|INPCK))
 
 static void tty0tty_set_termios(struct tty_struct *tty,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+                               const
+#endif
                                struct ktermios *old_termios)
 {
        unsigned int cflag;
```

**Comment by martinxyz** [comment]
Got this on Ubuntu 22, tested, the fix works.

**Comment by stappersg** [comment]
Issue #53 got closed by merging #52.

But the change from #52 is not that same change as this issue proposes.

Now we wait for further feedback ...

**Comment by stappersg** [comment]
Fixed in #57  (thanks #49)
